### PR TITLE
converted plan selector list and product to graphql

### PR DESCRIPTION
**Pull request by davidleger95**
## Reason for change

- convert plan selector list to graphql
- convert plan details (product data only) to graphql

TODO:
- update tests

## Testing

- check that plan selector works correctly in storybook

## Checklist

- [ ] **CHANGELOG**: The **Unreleased** section of CHANGELOG was updated
- [ ] **Prop changes**: [docs][docs] were updated
- [ ] **Prop changes**: E2E tests were updated, testing from the highest level possible
- [ ] **Platform testing**: If this change should be [tested against a platform][platform-testing], has it been?

[docs]: https://ui.sandbox.manifold.co
[platform-testing]: https://app.gitbook.com/@manifold/s/engineering/playbooks/testing-your-code-against-a-platform

## Audit

15 trials:

|                     | REST*         | GraphQL     |  Reduction |
|:------------|:------------|:------------|:-----------:|
| **Mean**    |  `184.9 ms` |   `133 ms`  |     29%        |
| **Median** | `186 ms`     |  `124 ms`    |    34%        |
| **Payload**|  `12.2 KB`   |   `2.7 KB`    |     78%        |

* when multiple queries are required in sequence, they are summed into total request time

### CODE REVIEW

File: src/components/manifold-plan-menu/manifold-plan-menu.tsx [hunk 1]
```diff
@@ -1,83 +1,102 @@
...
-            <div class="cost">
-              <manifold-skeleton-text>Free</manifold-skeleton-text>
-            </div>
-          </PlanButton>
-        ))}
-      </ul>
-    );
+    return <SkeletonPlanMenu />;
   }
 }
```
Comment (davidleger95) [inline src/components/manifold-plan-menu/manifold-plan-menu.tsx]: So I think this is a lot cleaner to use the class component as just the top level component and then use functional components to segment the UI into more specific pieces. Let me know if you have any thoughts or critiques of this approach.

File: src/components/manifold-plan-selector/manifold-plan-selector.tsx [hunk 2]
```diff
@@ -7,13 +8,95 @@ import { Marketplace } from '../../types/marketplace';
...
+                  dataCenter
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
```
Comment (davidleger95) [inline src/components/manifold-plan-selector/manifold-plan-selector.tsx]: Seems like an oversized query but it's only ~20ms slower (~100ms => ~120ms) to get all the plan features vs just the plan list fields. So this is more performant than breaking it up and lazy loading details when a plan is selected.

**Comment by davidleger95** [comment]
hmm, everythings working locally 🤔 @dangodev can you add me to the manifold team on zeit? not sure if i'm on there. I get a 404 when i try to look at the deployment details

File: src/components/manifold-plan-details/manifold-plan-details.spec.ts [hunk 3]
```diff
@@ -1,6 +1,15 @@
 import { Gateway } from '../../types/gateway';
 import { ManifoldPlanDetails as PlanDetails } from './manifold-plan-details';
-import { ExpandedPlan, ExpandedPlanCustom, Product } from '../../spec/mock/catalog';
+import { ExpandedPlan, ExpandedPlanCustom } from '../../spec/mock/catalog';
+import { Product } from '../../types/graphql';
+
+// GraphQL types are too strict and force adding unused fields here.
+const mockProduct: Product & any = {
```
Comment (dangodev) [inline src/components/manifold-plan-details/manifold-plan-details.spec.ts]: does `Partial<Product>` work better?

File: src/components/manifold-plan-menu/manifold-plan-menu.tsx (hunk 1 above)
Comment (dangodev) [reply src/components/manifold-plan-menu/manifold-plan-menu.tsx]: I’m happy with whatever we prefer as a group. I personally prefer the HTML being in one place personally, even if it’s verbose, but that’s not a strong opinion. Whatever makes sense to the most people! 🍌 

At any rate I think experiments like this are good.

File: src/components/manifold-plan-selector/manifold-plan-selector.spec.ts [hunk 4]
```diff
@@ -2,10 +2,21 @@ import { newSpecPage } from '@stencil/core/testing';
...
+      displayName: 'JawsDB MySQL',
+      label: 'jawsdb-mysql',
+      logoUrl: 'https://cdn.manifold.co/providers/jawsdb/logos/80ca8b9113cf76fd.png',
+    },
+  },
+};
+
 async function setup(productLabel?: string, resourceLabel?: string) {
   const page = await newSpecPage({
     components: [ManifoldPlanSelector],
```
Comment (dangodev) [inline src/components/manifold-plan-selector/manifold-plan-selector.spec.ts]: Protip: add `ManifoldPlanMenu`, etc. to this array if you’d like those components to be rendered as well. In many cases it’s helpful! Then you can do more E2E-style stuff, like test that a child component renders way down in the tree.

File: src/components/manifold-plan-details/manifold-plan-details.spec.ts (hunk 3 above)
Comment (davidleger95) [reply src/components/manifold-plan-details/manifold-plan-details.spec.ts]: I'll give it a try

**Comment by davidleger95** [comment]
> plans are no longer in ascending cost order in the left side menu

I was waiting to see if backend would pick this up but it seems like that's not a high priority at the moment, so I'll add that now.

I'll fix all the other stuff too :)

File: src/components/manifold-plan-selector/manifold-plan-selector.tsx [hunk 5]
```diff
@@ -7,12 +8,94 @@ import { Marketplace } from '../../types/marketplace';
...
 @Component({ tag: 'manifold-plan-selector' })
 export class ManifoldPlanSelector {
   @Element() el: HTMLElement;
   /** Show only free plans? */
-  @Prop() freePlans?: boolean = false;
+  @Prop() freePlans?: boolean;
+  /** _(hidden)_ Passed by `<manifold-connection>` */
+  @Prop() graphqlFetch?: GraphqlFetch = connection.graphqlFetch;
   /** _(hidden)_ Passed by `<manifold-connection>` */
   @Prop() restFetch?: RestFetch = connection.restFetch;
```
Comment (dangodev) [inline src/components/manifold-plan-selector/manifold-plan-selector.tsx]: We don’t need `restFetch` any more, do we?

File: src/components/manifold-plan-selector/manifold-plan-selector.tsx [hunk 6]
```diff
@@ -55,23 +138,18 @@ export class ManifoldPlanSelector {
...
+      query,
+      variables: { productLabel, free: this.freePlans },
     });
 
-    if (response && response.length) {
-      this.product = response[0]; // eslint-disable-line prefer-destructuring
-      await this.fetchPlans(this.product.id);
+    if (data && data.product) {
+      this.product = data.product;
+      this.fetchPlans(data.product.id);
```
Comment (dangodev) [inline src/components/manifold-plan-selector/manifold-plan-selector.tsx]: What’s in this query that’s not in the GQL response?

**Comment by davidleger95** [comment]
The metered pricing fixes require a lot of code changes. I think I'm going to create another branch off this one to keep the PR a reasonable size.

File: src/components/manifold-plan-selector/manifold-plan-selector.tsx (hunk 6 above)
Comment (davidleger95) [reply src/components/manifold-plan-selector/manifold-plan-selector.tsx]: There's nothing that isn't there. It's just a lot to convert at once. It's for the plan details (and I just realized it's in the cost component too)

**Comment by davidleger95** [comment]
>  Storybook: plans are no longer in ascending cost order in the left side menu

this is the only thing not handled at the moment. @Minivera is has a PR up for it

File: src/components/manifold-plan-menu/manifold-plan-menu.tsx [hunk 7]
```diff
@@ -1,83 +1,122 @@
...
+
 @Component({
   tag: 'manifold-plan-menu',
   styleUrl: 'plan-menu.css',
   shadow: true,
 })
 export class ManifoldPlanMenu {
-  @Prop() plans?: Catalog.ExpandedPlan[];
+  @Prop() plans?: PlanConnection;
+  @Prop() oldPlans: Catalog.ExpandedPlan[] = [];
```
Comment (davidleger95) [inline src/components/manifold-plan-menu/manifold-plan-menu.tsx]: feeding the old plans in as a temporary solution. Going to be converting this part over next.

File: src/components/manifold-resource-container/manifold-resource-container.tsx [hunk 8]
```diff
@@ -1,20 +1,41 @@
...
+  /** _(hidden)_ */
+  @Prop() graphqlFetch?: GraphqlFetch = connection.graphqlFetch;
   /** _(hidden)_ */
   @Prop() restFetch?: RestFetch = connection.restFetch;
   /** Which resource does this belong to? */
   @Prop() resourceLabel?: string;
   /** Set whether or not to refetch the resource from the api until it is in an available and valid state */
   @Prop() refetchUntilValid: boolean = false;
   @State() resource?: Gateway.Resource;
+  @State() gqlData?: Resource;
```
Comment (davidleger95) [inline src/components/manifold-resource-container/manifold-resource-container.tsx]: this is temporary. going to rename once everything is converted to gql

**Comment by dangodev** [comment]
Looking good! I know you’ll get the plan order sorted out, but one last little nitpick:

When loading the plans, they all flash “free” for a brief moment while loading. I previously had a check in place for that, where it’d treat `cost = undefined` differently than `cost = 0` (free). If you go back and forth between the plan types in Storybook, it’s apparent.

**Pull request merged**